### ISSUE TITLE
File handles are treated differently if file is mounted from Windows host

### ISSUE THREAD

**Issue by liayn** [moby/moby#39136]
**Description**

I have a sample PHP script (see below) that shows different behaviour depending on the file location of the file opened from within the script.

The general issue is that a process is unable to open a file a second time if the file has been deleted meanwhile, IF the file is located in a bind-mount from the Windows host.
Running the same script with a file location inside the container (or generally on any Linux host), no problems occur.

I found this issue while investigating a bug in TYPO3. See bug report with all the details here:
https://forge.typo3.org/issues/88197

Thank you very much for any input or action on how to make it possible to have Docker for Windows working the same way as Docker on Linux.

```
<?php
$fileName = '/some/host/mounted/location/foo.txt'; // this fails
// $fileName = '/tmp/foo.txt'; // this works

$h = fopen($fileName, 'c');
if (!$h) echo "went wrong";

unlink($fileName);

$h = fopen($fileName, 'c');
if (!$h) echo "went wrong too";

echo "the end";
```

**Steps to reproduce the issue:**
1. create a docker-container from e.g. php7.2-apache with a bind-mount from the Windows Host machine. (I assume '/some/host/mounted/location/' to be the mounted path)
2. create a php file with the content above in the docker-container (`/var/www/html/test.php`)
3. access the container with your browser (http://localhost/test.php)
4. See the error message

Change the first line to the working one and repeat step 3. You will see no error message only the text "the end" in the browser.

**Describe the results you received:**

Error Message:
```
PHP Warning: fopen(foo.txt): failed to open stream: No such file or directory in /var/www/html/test.php
```

**Describe the results you expected:**

No error message at all.

**Additional information you deem important (e.g. issue happens only occasionally):**

This is reproducable on Docker for Windows only.
While I'm aware that you should close the opened file before you open it again, the current behaviour of PHP (or Linux?) is rather forgiving in this regard and the code works flawlessly on any native Linux machine or on native file systems.

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           18.09.2
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        6247962
 Built:             Sun Feb 10 04:12:31 2019
 OS/Arch:           windows/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.2
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.6
  Git commit:       6247962
  Built:            Sun Feb 10 04:13:06 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Output of `docker info`:**

```
Containers: 114
 Running: 2
 Paused: 0
 Stopped: 112
Images: 407
Server Version: 18.09.2
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9754871865f7fe2f4e74d43e2fc7ccd237edcbce
runc version: 09c8266bf2fcf9519a651b04ae54c967b9ab86ec
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.125-linuxkit
Operating System: Docker for Windows
OSType: linux
Architecture: x86_64
CPUs: 3
Total Memory: 1.934GiB
Name: linuxkit-00155d486c33
ID: T62F:5GXM:R64A:NOVW:6OPS:TEON:A5Q5:27BV:RKJT:OSGK:WI6G:5QTP
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 38
 Goroutines: 59
 System Time: 2019-04-24T19:27:06.0753418Z
 EventsListeners: 1
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by thaJeztah** [comment]
If you `docker exec` into the container, is the file visible?

File sharing between the Windows host and the VM in which the daemon (and container) runs on docker for Windows is using CIFS, so there might be some differences (in addition to permissions not really mapping between Windows and Linux).

Wondering in this case if there's either some delay, or if file sharing failed altogether

**Comment by liayn** [comment]
Thanks for your reply @thaJeztah. The file is really gone after the `unlink` call. Also if you look inside container. So for me this really looks like an implementation issue at a lower level.

Specifically keep in mind that the `fopen(.., 'c')` in PHP is described with

> Open the file for writing only. If the file does not exist, it is created.

**Comment by thaJeztah** [comment]
Does it make a difference if you `fclose($h)` before you `unlink($filename)`? Also might be worth checking the return value of `unlink()` to check if the delete was successful.
I'm not on Windows myself, so can't test the situation easily.

Perhaps it could still be a race condition, where the `unlink()` takes some time, so the second `fopen()` detects that the file exists (so doesn't create it), but at the same time it's being removed


@StefanScherer @guillaumerose any ideas?

**Comment by liayn** [comment]
Yes, closing the handle correctly makes the whole story work. It does not matter whether you close before or after the `unlink` call.

**Comment by thaJeztah** [comment]
I think cifs uses file locking, in which case the `unlink` and/or re-creating of the file would fail if the file is still locked.

**Comment by liayn** [comment]
The `unlink` works as I know the file is gone on the host
